gobblin-data-management/src/main/java/org/apache/gobblin/data/management/copy/CopyableFile.java

```diff
@@ -354,6 +359,36 @@ public static OwnerAndPermission resolveReplicatedOwnerAndPermission(FileSystem
     return ownerAndPermissions;
   }
 
+  public static Map<String, OwnerAndPermission> resolveReplicatedAncestorOwnerAndPermissionsRecursively(FileSystem sourceFs, Path fromPath,
+      Path toPath, CopyConfiguration copyConfiguration) throws IOException {
+
+    Map<String, OwnerAndPermission> ownerAndPermissions = Maps.newHashMap();
+
+    // We only pass directories to this method anyways. Those directories themselves need permissions set.
+    Path currentOriginPath = fromPath;
+    Path currentTargetPath = toPath;
+
+    if (!PathUtils.isAncestor(currentTargetPath, currentOriginPath)) {
+      throw new IOException(String.format("currentTargetPath %s must be an ancestor of currentOriginPath %s.", currentTargetPath, currentOriginPath));
+    }
+
+    while (PathUtils.isAncestor(currentTargetPath, currentOriginPath.getParent())) {
+      ownerAndPermissions.put(PathUtils.getPathWithoutSchemeAndAuthority(currentOriginPath).toString(), resolveReplicatedOwnerAndPermission(sourceFs, currentOriginPath, copyConfiguration));
+      currentOriginPath = currentOriginPath.getParent();
+    }
+    // Now currentTargetPath and currentOriginPath are the same path.
+
+    // Walk through the parents and preserve the permissions from Origin -> Target as we go in lockstep.
+    while (currentOriginPath != null && currentTargetPath != null
```

Comment:
What's the reason that we want to change the parent permission as well

Reply:
Sometimes user just wants to.
e.g. if source dir structure is /d1/d2/dataset/partition/hourly/2022/01/25/10
and they want to copy /d1/d2/dataset/partition/hourly/2022/01/25 and still wants to replicate permissions on  /d1/d2/dataset/partition,  /d1/d2/dataset/partition/hourly,  /d1/d2/dataset/partition/hourly,  /d1/d2/dataset/partition/hourly/2022,  /d1/d2/dataset/partition/hourly/2022/01 then this option will be needed

Reply:
two questions:
1. should we make this behavior configurable as one add-on option but not the default behavior?
2. in the example you give, we will even set the permission of /d1 and /d1/d2 to be the same, not sure will that cause issue

Reply:
1. For now, this logic can be used with config USE_NEW_PRESERVE_LOGIC_KEY
2. Yes, it will set permissions on /d1 and /d1/d2 as well. If user thinks that's an issue, he should not choose this logic. At this point, there is no friendly way to tell exactly upto which ancestors permissions need to be applied. I have provided a config 'stopOnError' which would stop setting permissions if it encounters any error